proxy/http2/Http2ConnectionState.cc

```diff
@@ -242,7 +242,7 @@ rcv_headers_frame(Http2ConnectionState &cstate, const Http2Frame &frame)
   if (cstate.is_valid_streamid(stream_id)) {
     stream = cstate.find_stream(stream_id);
     if (stream == nullptr || !stream->has_trailing_header()) {
-      return Http2Error(Http2ErrorClass::HTTP2_ERROR_CLASS_CONNECTION, Http2ErrorCode::HTTP2_ERROR_STREAM_CLOSED,
```

Comment:
Is this for `http2/6.2/2` of h2spec?

5.1.  Stream States says below

> closed:
>      An endpoint that receives any frame other than PRIORITY
>      after receiving a RST_STREAM MUST treat that as a stream error
>      (Section 5.4.2) of type STREAM_CLOSED.

Probably, we need to break down the conditions in line 244
- `stream == nullptr` -> `STREAM_CLOSED`
- `!stream->has_trailing_header()` -> HTTP2_ERROR_PROTOCOL_ERROR

Reply:
This was to fix one of the h2spec cases.  I'll update to split the cases here to return two different error codes.

Reply:
Specifically it was a problem in the 8 section.

```
Hypertext Transfer Protocol Version 2 (HTTP/2)
               8. HTTP Message Exchanges
                 8.1. HTTP Request/Response Exchange
                   × 1: Sends a second HEADERS frame without the END_STREAM flag
                     -> The endpoint MUST respond with a stream error of type PROTOCOL_ERROR.
                        Expected: GOAWAY Frame (Error Code: PROTOCOL_ERROR)
                                  RST_STREAM Frame (Error Code: PROTOCOL_ERROR)
                                  Connection closed
                          Actual: Timeout
             
                   8.1.2. HTTP Header Fields
                     8.1.2.1. Pseudo-Header Fields
                       × 3: Sends a HEADERS frame that contains a pseudo-header field as trailers
                         -> The endpoint MUST respond with a stream error of type PROTOCOL_ERROR.
                            Expected: GOAWAY Frame (Error Code: PROTOCOL_ERROR)
                                      RST_STREAM Frame (Error Code: PROTOCOL_ERROR)
                                      Connection closed
                              Actual: Timeout
```

I assume in earlier version of the PR when the netvc close was in the Http2ClientSession::do_io_close() the "Connection closed" case triggered and addressed these cases.

Splitting the checks as you suggest also solves the issue.